File: android/app/src/main/java/com/familyrecipe/core/models/Family.kt
package com.familyrecipe.core.models

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverter
import androidx.room.TypeConverters
import java.util.Date
import java.util.UUID

// MARK: - Family Model

@Entity(tableName = "families")
@TypeConverters(FamilyConverters::class)
data class Family(
    @PrimaryKey
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val familyDescription: String = "",
    val inviteCode: String = generateInviteCode(),
    val adminMemberId: String,
    val memberIds: List<String> = listOf(),
    val theme: TemplateKey = TemplateKey.VINTAGE,
    val language: LanguageCode = LanguageCode.EN,
    val units: UnitSystem = UnitSystem.IMPERIAL,
    val iconEmoji: String = "\uD83D\uDC68\u200D\uD83D\uDC69\u200D\uD83D\uDC67\u200D\uD83D\uDC66", // Family emoji
    val createdAt: Date = Date(),
    val lastActivityAt: Date = Date()
) {
    val memberCount: Int
        get() = memberIds.size

    val canAddMembers: Boolean
        get() = memberIds.size < 20

    fun withNewInviteCode(): Family = copy(
        inviteCode = generateInviteCode(),
        lastActivityAt = Date()
    )

    fun withNewMember(memberId: String): Family? {
        if (!canAddMembers || memberIds.contains(memberId)) return null
        return copy(
            memberIds = memberIds + memberId,
            lastActivityAt = Date()
        )
    }

    fun withoutMember(memberId: String): Family? {
        if (memberId == adminMemberId) return null // Can't remove admin
        return copy(
            memberIds = memberIds.filter { it != memberId },
            lastActivityAt = Date()
        )
    }

    fun isAdmin(memberId: String): Boolean = memberId == adminMemberId

    fun withTransferredAdmin(newAdminId: String): Family? {
        if (!memberIds.contains(newAdminId)) return null
        return copy(
            adminMemberId = newAdminId,
            lastActivityAt = Date()
        )
    }

    fun withUpdatedActivity(): Family = copy(lastActivityAt = Date())

    companion object {
        private const val INVITE_CODE_CHARACTERS = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

        fun generateInviteCode(): String {
            return (1..6)
                .map { INVITE_CODE_CHARACTERS.random() }
                .joinToString("")
        }
    }
}

// MARK: - Enums

enum class TemplateKey(val displayName: String, val description: String) {
    VINTAGE(
        displayName = "Vintage Cookbook",
        description = "Warm, nostalgic, like grandma's recipe cards"
    ),
    MODERN(
        displayName = "Modern Kitchen",
        description = "Clean, minimal, contemporary design"
    ),
    PLAYFUL(
        displayName = "Playful Family",
        description = "Fun, colorful, perfect for cooking with kids"
    )
}

enum class LanguageCode(val displayName: String) {
    EN("English"),
    ES("Español"),
    HI("हिन्दी")
}

enum class UnitSystem(val displayName: String) {
    IMPERIAL("Imperial (cups, °F)"),
    METRIC("Metric (ml, °C)")
}

// MARK: - Room Type Converters

class FamilyConverters {
    private val gson = com.google.gson.Gson()

    @TypeConverter
    fun fromStringList(value: List<String>): String = gson.toJson(value)

    @TypeConverter
    fun toStringList(value: String): List<String> {
        val type = object : com.google.gson.reflect.TypeToken<List<String>>() {}.type
        return gson.fromJson(value, type) ?: emptyList()
    }

    @TypeConverter
    fun fromTemplateKey(value: TemplateKey): String = value.name

    @TypeConverter
    fun toTemplateKey(value: String): TemplateKey = TemplateKey.valueOf(value)

    @TypeConverter
    fun fromLanguageCode(value: LanguageCode): String = value.name

    @TypeConverter
    fun toLanguageCode(value: String): LanguageCode = LanguageCode.valueOf(value)

    @TypeConverter
    fun fromUnitSystem(value: UnitSystem): String = value.name

    @TypeConverter
    fun toUnitSystem(value: String): UnitSystem = UnitSystem.valueOf(value)

    @TypeConverter
    fun fromDate(value: Date): Long = value.time

    @TypeConverter
    fun toDate(value: Long): Date = Date(value)
}

// MARK: - Sample Data

object FamilySamples {
    fun sampleFamily(): Pair<Family, FamilyMember> {
        val memberId = UUID.randomUUID().toString()
        val familyId = UUID.randomUUID().toString()

        val family = Family(
            id = familyId,
            name = "The Recipe Testers",
            familyDescription = "A family that loves to cook together!",
            adminMemberId = memberId,
            memberIds = listOf(memberId),
            theme = TemplateKey.VINTAGE,
            language = LanguageCode.EN,
            units = UnitSystem.IMPERIAL,
            iconEmoji = "\uD83D\uDC68\u200D\uD83D\uDC69\u200D\uD83D\uDC67\u200D\uD83D\uDC66"
        )

        val member = FamilyMember(
            id = memberId,
            name = "Chef",
            avatarEmoji = "\uD83D\uDC68\u200D\uD83C\uDF73",
            role = FamilyRole.ADMIN,
            familyId = familyId
        )

        return Pair(family, member)
    }
}

File: ios/Core/Models/Family.swift
import Foundation
import SwiftData

// MARK: - Family Model

@Model
final class Family {
    @Attribute(.unique) var id: UUID
    var name: String
    var familyDescription: String
    var inviteCode: String
    var adminMemberId: UUID
    var memberIds: [UUID]

    // Customization
    var theme: TemplateKey
    var language: LanguageCode
    var units: UnitSystem
    var iconEmoji: String

    // Timestamps
    var createdAt: Date
    var lastActivityAt: Date

    // Computed
    var memberCount: Int {
        memberIds.count
    }

    var canAddMembers: Bool {
        memberIds.count < 20
    }

    init(
        id: UUID = UUID(),
        name: String,
        familyDescription: String = "",
        inviteCode: String? = nil,
        adminMemberId: UUID,
        memberIds: [UUID]? = nil,
        theme: TemplateKey = .vintage,
        language: LanguageCode = .en,
        units: UnitSystem = .imperial,
        iconEmoji: String = "👨‍👩‍👧‍👦",
        createdAt: Date = Date(),
        lastActivityAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.familyDescription = familyDescription
        self.inviteCode = inviteCode ?? Family.generateInviteCode()
        self.adminMemberId = adminMemberId
        self.memberIds = memberIds ?? [adminMemberId]
        self.theme = theme
        self.language = language
        self.units = units
        self.iconEmoji = iconEmoji
        self.createdAt = createdAt
        self.lastActivityAt = lastActivityAt
    }

    // MARK: - Methods

    static func generateInviteCode() -> String {
        let characters = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789" // Excluding similar chars (0/O, 1/I)
        return String((0..<6).map { _ in characters.randomElement()! })
    }

    func regenerateInviteCode() {
        inviteCode = Family.generateInviteCode()
        lastActivityAt = Date()
    }

    func addMember(_ memberId: UUID) -> Bool {
        guard canAddMembers, !memberIds.contains(memberId) else { return false }
        memberIds.append(memberId)
        lastActivityAt = Date()
        return true
    }

    func removeMember(_ memberId: UUID) -> Bool {
        guard memberId != adminMemberId else { return false } // Can't remove admin
        memberIds.removeAll { $0 == memberId }
        lastActivityAt = Date()
        return true
    }

    func isAdmin(_ memberId: UUID) -> Bool {
        memberId == adminMemberId
    }

    func transferAdmin(to newAdminId: UUID) -> Bool {
        guard memberIds.contains(newAdminId) else { return false }
        adminMemberId = newAdminId
        lastActivityAt = Date()
        return true
    }

    func updateActivity() {
        lastActivityAt = Date()
    }
}

// MARK: - Enums (TemplateKey moved to TemplateTokens.swift)

enum LanguageCode: String, Codable, CaseIterable {
    case en
    case es
    case hi

    var displayName: String {
        switch self {
        case .en: return "English"
        case .es: return "Español"
        case .hi: return "हिन्दी"
        }
    }
}

enum UnitSystem: String, Codable, CaseIterable {
    case imperial
    case metric

    var displayName: String {
        switch self {
        case .imperial: return "Imperial (cups, °F)"
        case .metric: return "Metric (ml, °C)"
        }
    }
}

// MARK: - Sample Data

extension Family {
    static func sampleFamily() -> (family: Family, member: FamilyMember) {
        let memberId = UUID()
        let familyId = UUID()

        let family = Family(
            id: familyId,
            name: "The Recipe Testers",
            familyDescription: "A family that loves to cook together!",
            adminMemberId: memberId,
            memberIds: [memberId],
            theme: .vintage,
            language: .en,
            units: .imperial,
            iconEmoji: "👨‍👩‍👧‍👦"
        )

        let member = FamilyMember(
            id: memberId,
            name: "Chef",
            avatarEmoji: "👨‍🍳",
            role: .admin,
            familyId: familyId
        )

        return (family, member)
    }
}
